bench/wtperf/wtperf.c

```diff
@@ -1386,24 +1396,42 @@ flush_tier_worker(void *arg)
         goto err;
     }
 
-    while (!wtperf->stop) {
+    stop = false;
+    while (!stop) {
         /* Break the sleep up, so we notice interrupts faster. */
         for (i = 0; i < opts->tiered_flush_interval; i++) {
             sleep(1);
-            if (wtperf->stop)
+            /*
+             * We need to know if we stop before the call to flush_tier. Don't keep rereading the
+             * global value.
+             */
+            stop = wtperf->stop;
+            if (stop || wtperf->error)
                 break;
         }
-        /* If the workers are done, don't bother with a final call. */
-        if (wtperf->stop)
+        /* If workers are done, do a final call to flush that last data. */
+        if (wtperf->error)
             break;
-
+        /*
+         * In order to get all the data into the last object when the work is done, we need to call
+         * checkpoint to get all the data into this object before calling flush_tier.
+         */
+        if (stop) {
+            lprintf(wtperf, 0, 1, "Last call before stopping flush_tier");
+            if ((ret = session->checkpoint(session, NULL)) != 0) {
```

Comment:
Just an observation that, when stopping, this thread does a checkpoint, flush_tier, while the checkpoint thread will do a checkpoint.  So the final sequence may be "checkpoint, checkpoint, flush_tier" or "checkpoint, flush_tier, checkpoint" depending on the luck of the draw.

Reply:
It shouldn't be because the checkpoint thread stops based on `wtperf->ckpt_stop` not `wtperf->stop` now. We set `wtperf->stop` first and then stop any flush tier threads and then stop the checkpoint thread. So that means we will always have a checkpoint after the `flush_tier`. We could, as you indicate, end up with two checkpoints before the flush tier and that is okay. The flushing thread wants to guarantee the checkpoint happens when it is stopping to push the data into the object.

Reply:
Thanks, nice to know it is more predictable.